### ISSUE TITLE
Federation manual testing on development branch: lags and instability

### ISSUE THREAD

**Issue by cmrd-senya** [diaspora/diaspora#6482]
I ran two VMs with Diaspora pods to do some testing for #5522.

While usual posts and people federated just fine, I experienced lags with likes, comments, notifications and sharing federation.

When I clicked "Like" link below any post the both applications started to process an event. Then it hanged several times for 25 seconds with the last line in the log "DEBUG PID-3197 TID-18831720 Rails: ETHON: started MULTI". After several such hangs (like it was an operation time out), it returned back to normal operation. The like either was posted or was not posted - it was quite random.

At the same time I had no problems with seeing post across the pods immediately after they were written.

Here are the logs of the servers (from the moment of the event start):
[The pod who liked](https://gist.githubusercontent.com/cmrd-senya/4d45c41d0b8313e98653/raw/8fa8eb73175613972123c65fbff4e21b00c0f9d3/pod-who-liked)
[The pod who was liked](https://gist.githubusercontent.com/cmrd-senya/4d45c41d0b8313e98653/raw/8fa8eb73175613972123c65fbff4e21b00c0f9d3/pod-who-was-liked)

**Comment by jhass** [comment]
Well, it did time out, not sure there's anything we can do about

```

[2015-10-10T04:50:23] WARN  PID-2815 TID-18831720 HydraWrapper: event=http_multi_fail sender_id=1 url=http://192.168.56.102/receive/public return_code=operation_timedout response_code=0
[2015-10-10T04:50:48] DEBUG PID-2815 TID-18831720 Rails: ETHON:         performed EASY effective_url=http://192.168.56.102/receive/public response_code=0 return_code=operation_timedout total_time=0.200405
[2015-10-10T04:51:13] DEBUG PID-2815 TID-18831720 Rails: ETHON:         performed EASY effective_url=http://192.168.56.102/receive/public response_code=0 return_code=operation_timedout total_time=0.202368
```
```
[2015-10-10T04:51:13] DEBUG PID-4643 TID-3996000 Rails: ETHON:         performed EASY effective_url=http://192.168.56.103/receive/public response_code=0 return_code=operation_timedout total_time=0.200404

**Comment by jhass** [comment]
https://github.com/diaspora/diaspora/blob/develop/config/diaspora.yml.example#L507-L510 if you want more output.

**Comment by cmrd-senya** [comment]
With this setting I see
```
[2015-10-12T21:35:38] DEBUG PID-4798 TID-5634400 Rails: ETHON: started MULTI
Hostname was NOT found in DNS cache
  Trying 192.168.56.103...
Connected to 192.168.56.103 (192.168.56.103) port 80 (#0)
POST /receive/users/3815bfe0533b0133a4080800276d7a8d HTTP/1.1
Host: 192.168.56.103
Accept: */*
User-Agent: Diaspora 0.5.99.0-p09f4eeac
Content-Length: 4740
Content-Type: application/x-www-form-urlencoded

upload completely sent off: 4740 out of 4740 bytes
```

Everything else is same.

I need to debug, why it answers on some types of requests, and timeouts on others.

**Comment by cmrd-senya** [comment]
@jhass, do you have any idea, how is it possible, that it works differently depending on post types?

**Comment by jhass** [comment]
@cmrd-senya I think what's happening is that the remote pod is trying to relay the like back to the sending pod and since you run both in development mode it tries that in the request the sending pod is sending. Now the sending pod is doing its sending in request too and we're only running one unicorn worker in development mode, so it's blocked and can't accept the the request from the remote pod relaying the post. A deadlock occurs and the timeout is hit.

Try setting https://github.com/diaspora/diaspora/blob/develop/config/diaspora.yml.example#L184

**Comment by cmrd-senya** [comment]
Yes, that helped, thanks!

I also noticed what I think is an issue and I think it should be known, but I didn't find it on the tracker.

Given we have (freshly registered) users:
alice@pod1.local
bob@pod2.local
carol@pod2.local

Alice shares with Bob, so the pods know about each other.

Carol post a public status message. We check the Carol's page from the Alice's page. Pod1 fetches the profile of Carol from Pod2 and her public message along with the profile. Then Carol posts a new public status message. We refresh Carol's page from Alice's page on Pod1, and we **don't see** the new Carol's post.

Now if we setup connection between Carol and Bob, and then login to the Alice's account on Pod1 and to the Carol's account on Pod2, and if we make a public post from the Carol's account, the we **will see* it from Alice's page.

**So, to sum up**: If two pods have a connection and a user on one of the pods posts public status, it gets federated to the other pod only if the user is connected to anyone on the other pod.

If this issue is not registered yet, I could report it separately.

**Comment by jhass** [comment]
How do you define "two pods have a connection"? Currently it's defined as "a user shares with/follows a user of the other pod" and if I got it right what you describe is very much by design.

**Comment by cmrd-senya** [comment]
I mean pods are informed on existing of each other. If I'm not mistaken, there is a table in the database where known pods are listed?

Ok, if it is how it meant to work. I just thought that if a post is public, then it should be federated to all of the known pods, not only to those where we have "friends".

**Comment by cmrd-senya** [comment]
Ok, anyway, this one is solved.
Thanks, @jhass.

**Comment by Flaburgan** [comment]
To not send a post to a pod (even known) if the user doesn't have a contact there is the normal behaviour. We would be flooding the whole network otherwise.
